File: merge_sort_inversions/merge_sort_inversions.cpp
#include <iostream>
#include <vector>
#include <fstream>

using namespace std;

class Solution {
public:
    using VL = vector<long>;
    using Pair = pair<VL, long>;
    using fun = function<Pair(VL&&)>;
    Pair merge(VL& A, VL& B, VL C = {}, long inv = 0) {
        auto i = 0,
             j = 0;
        while (i < A.size() && j < B.size()) {
            if (A[i] < B[j]) {
                C.push_back(A[i++]);
            } else {
                inv += A.size() - i; // ⭐️ B[j] comes before all remaining A[i...], thus all remaining A[i...] are inversions
                C.push_back(B[j++]);
            }
        }
        C.insert(C.end(), A.begin() + i, A.end());
        C.insert(C.end(), B.begin() + j, B.end());
        return { C, inv };
    }
    Pair inversions(VL& A) {
        fun go = [&](VL&& A) -> Pair {
            int N = A.size();
            if (N < 2)
                return { A, 0 };
            int half = N / 2;
            auto [first,  inv1] = go({ A.begin(), A.begin() + half });
            auto [second, inv2] = go({ A.begin() + half, A.end() });
            auto [third,  inv3] = merge(first, second);
            return { third, inv1 + inv2 + inv3 };
        };
        return go(move(A));
    }
};

long run(string filename) {
    Solution solution;
    Solution::VL A;
    fstream fin{ filename };
    for (string line; fin >> line; A.push_back(stol(line)));
    auto [_, inv] = solution.inversions(A);
    return inv;
}

int main() {
    cout << "problem3.5test.txt: " << run("problem3.5test.txt") << endl   // problem3.5test.txt: 28
         << "problem3.5.txt: "     << run("problem3.5.txt")     << endl;  // problem3.5.txt: 2407905288
    return 0;
}

File: merge_sort_inversions/merge_sort_inversions.jl
#
# Merge Sort Inversions
#
# Input: array A of n distinct integers
# Output: the number of inversions of A
#
function go(A)
  N = length(A)
  if N < 2
    return A, 0
  end
  half = div(N, 2)
  first, inv1 = go(A[begin:half])
  second, inv2 = go(A[half+1:end])
  third, inv3 = merge(first, second)
  return third, inv1 + inv2 + inv3
end

function merge(A, B)
  C, inv = [], 0
  i, j, k = 1, 1, 1
  while i <= length(A) && j <= length(B)
    if A[i] < B[j]
      push!(C, A[i])
      i += 1
    else
      push!(C, B[j]); inv += length(A) - i + 1  # ⭐️ B[j] comes before all remaining A[i...], thus all remaining A[i...] are inversions
      j += 1
    end
    k += 1
  end
  append!(C, @view A[i:end])
  append!(C, @view B[j:end])
  return C, inv
end

function run(filename)
  input = Int[]
  open(filename, "r") do file
    for line in eachline(file)
      push!(input, parse(Int, strip(line)))
    end
  end
  _, inv = go(input)
  return inv
end

println(run("problem3.5test.txt"))  # 28
println(run("problem3.5.txt"))      # 2407905288
